[LV] Enable auto-vectorisation of loops with uncountable exits

david-arm opened the pull request:
Until now the feature to enable vectorisation of some early exit
loops with uncountable exits was controlled under a flag, off by
default. Now that we have efficient code generation for
vectorising such loops (see PR #130766) and we are still far
enough away from the next LLVM release it seems like a good time
to enable the feature by default. If any issues arise
post-commit it can be easily reverted.

Using this patch I built and ran the LLVM test suite successfully,
which on neoverse-v1 led to the vectorisation of 114 additional
early exit loops. I then performed a bootstrap build of clang,
which built cleanly with 64 extra early exit loops vectorising.

I also built and ran SPEC2017 successfully with no change in
performance for both neoverse-v1 and neoverse-v2.

NOTE: This is currently blocked on #128061

fhahn reviewed (commented):
I think it would be good to fix https://github.com/llvm/llvm-project/pull/128061 before enabling, as without it we currently may compute incorrect bounds for early-exit loops, which in turn are used in `isDereferenceableAndAlignedInLoop`

david-arm commented:
> I think it would be good to fix #128061 before enabling, as without it we currently may compute incorrect bounds for early-exit loops, which in turn are used in `isDereferenceableAndAlignedInLoop`

Oh yeah sure, that makes absolute sense. That's on my list of PRs to review today!

david-arm commented:
> I tried this out on RISC-V and we get 107% more loops vectorized on one of the SPEC CPU 2017 benchmarks, with an overall geomean 5% improvement across all the benchmarks.
> 
> There's also an LNT run [here](https://lnt.lukelau.me/db_default/v4/nts/369?show_delta=yes&show_previous=yes&show_stddev=yes&show_mad=yes&show_all=yes&show_all_samples=yes&show_sample_counts=yes&show_small_diff=yes&num_comparison_runs=0&test_filter=&test_min_value_filter=&aggregation_fn=min&MW_confidence_lv=0.05&compare_to=370&submit=Update) and there's no regressions. So no issues on the RISC-V side!

OK great! I'm glad that you see no regressions for RISC-V, although it's worth saying this patch only starts vectorising early exit loops if we know something about the object we're loading from. For example, if you can prove that all loads are within the bounds of an `alloca` then it's safe to vectorise.

fhahn reviewed (commented):
> I just started looking at early exits again after our initial attempt at it a while ago, also on llvm-dev channel: https://github.com/annamthomas/work-notes/blob/main/MultiDataDependentExitVectorization.rst).
> 
> @fhahn @david-arm, are there any other known functional issues blocking the early-exit option from being switched on by default apart from #128061?
> 

That sounds right, should have the patch updated by end of week.

> I'll run some tests downstream with the option ON and see what falls out.

sounds great, thanks

david-arm commented:
> I just started looking at early exits again after our initial attempt at it a while ago, also on llvm-dev channel: https://github.com/annamthomas/work-notes/blob/main/MultiDataDependentExitVectorization.rst).
> 
> @fhahn @david-arm, are there any other known functional issues blocking the early-exit option from being switched on by default apart from #128061?
> 
> I'll run some tests downstream with the option ON and see what falls out.

No issues that I know of. I was just waiting for https://github.com/llvm/llvm-project/pull/128061. I can also run some downstream tests again and rebase this patch!

david-arm committed:
[LV] Enable auto-vectorisation of loops with uncountable exits

Until now the feature to enable vectorisation of some early exit
loops with uncountable exits was controlled under a flag, off by
default. Now that we have efficient code generation for
vectorising such loops (see PR #130766) and we are still far
enough away from the next LLVM release it seems like a good time
to enable the feature by default. If any issues arise
post-commit it can be easily reverted.

Using this patch I built and ran the LLVM test suite successfully,
which on neoverse-v1 led to the vectorisation of 114 additional
early exit loops. I then performed a bootstrap build of clang,
which built cleanly with 64 extra early exit loops vectorising.

I also built and ran SPEC2017 successfully with no change in
performance for both neoverse-v1 and neoverse-v2.

david-arm commented:
Rebased this patch now that https://github.com/llvm/llvm-project/pull/128061 has landed. LLVM test suite seems fine, but I'll run some more tests to verify.

annamthomas commented:
We've run our fuzzer downstream for couple of days with dereferencability checks disabled (to catch more early exit loops) and stress test the vectorizer with -force-vector-width. One miscompile was found, but that's not related to early-exit vectorization. 
LGTM from me.

fhahn reviewed (commented):
I ran into some additional mis-compiles when testing, in case interleaving is forced (e.g. via pragmas).

AnyOf (https://github.com/llvm/llvm-project/pull/145340), FirstActiveLane (https://github.com/llvm/llvm-project/pull/145394) and ExtractElement vec (FirstActiveLane)) need to be handled when unrolling. 

End-to-end tests are here: https://github.com/llvm/llvm-test-suite/pull/264

fhahn reviewed (commented):
> Is there a way to proceed with this patch without making it dependent on interleave pragma support. Given the pragma is really just a hint, we could just ignore the hint?

Sure that should be fine. It should prevent interleaving in any case, i.e. also when forced via options.

Although I don't think getting interleaving fixed will take too long.

david-arm committed:
Merge branch 'main' into ee_enable

fhahn reviewed (approved):
LGTM, with https://github.com/llvm/llvm-project/pull/145877 merged, thanks

bgra8 commented:
@david-arm please note that we have bisected a bunch of `clang` crashes at this revision:

```
Stack dump:
0.      Program arguments: /bin/clang --target=x86_64-unknown-linux-gnu -O3 -c -x ir -mllvm -force-vector-interleave=2 -o /dev/null reduced.ll
1.      Optimizer
2.      Running pass "function<eager-inv>(float2int,lower-constant-intrinsics,chr,loop(loop-rotate<header-duplication;no-prepare-for-lto>,loop-deletion),loop-distribute,inject-tli-mappings,loop-vectorize<no-interleave-forced-only;no-vectorize-forced-only;>,infer-alignment,loop-load-elim,instcombine<max-iterations=1;no-verify-fixpoint>,simplifycfg<bonus-inst-threshold=1;forward-switch-cond;switch-range-to-icmp;switch-to-lookup;no-keep-loops;hoist-common-insts;no-hoist-loads-stores-with-cond-faulting;sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,slp-vectorizer,vector-combine,instcombine<max-iterations=1;no-verify-fixpoint>,loop-unroll<O3>,transform-warning,sroa<preserve-cfg>,infer-alignment,instcombine<max-iterations=1;no-verify-fixpoint>,loop-mssa(licm<allowspeculation>),alignment-from-assumptions,loop-sink,instsimplify,div-rem-pairs,tailcallelim,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;hoist-loads-stores-with-cond-faulting;no-sink-common-insts;speculate-blocks;simplify-cond-branch;speculate-unpredictables>)" on module "reduced.ll"
3.      Running pass "loop-vectorize<no-interleave-forced-only;no-vectorize-forced-only;>" on function "barney"
 #0 0x000055888889874a llvm::sys::RunSignalHandlers() (/bin/clang+0x889874a)
 #1 0x000055888885d828 CrashRecoverySignalHandler(int) (/bin/clang+0x885d828)
 #2 0x00007ff9ded99e80 __restore_rt (/lib64/libpthread.so.0+0x14e80)
 #3 0x00005588886f037d llvm::FunctionType::get(llvm::Type*, llvm::ArrayRef<llvm::Type*>, bool) (/bin/clang+0x86f037d)
 #4 0x000055888868706b llvm::Intrinsic::getType(llvm::LLVMContext&, unsigned int, llvm::ArrayRef<llvm::Type*>) (/bin/clang+0x868706b)
 #5 0x000055888868dc45 llvm::Intrinsic::getOrInsertDeclaration(llvm::Module*, unsigned int, llvm::ArrayRef<llvm::Type*>) (/bin/clang+0x868dc45)
 #6 0x0000558888667b00 llvm::IRBuilderBase::getReductionIntrinsic(unsigned int, llvm::Value*) (/bin/clang+0x8667b00)
 #7 0x0000558887bf44ac llvm::VPInstruction::generate(llvm::VPTransformState&) (/bin/clang+0x7bf44ac)
 #8 0x0000558887bf5b35 llvm::VPInstruction::execute(llvm::VPTransformState&) (/bin/clang+0x7bf5b35)
 #9 0x0000558887bd47de llvm::VPBasicBlock::execute(llvm::VPTransformState*) (/bin/clang+0x7bd47de)
#10 0x0000558887bd830e llvm::VPlan::execute(llvm::VPTransformState*) (/bin/clang+0x7bd830e)
#11 0x0000558887abd126 llvm::LoopVectorizationPlanner::executePlan(llvm::ElementCount, unsigned int, llvm::VPlan&, llvm::InnerLoopVectorizer&, llvm::DominatorTree*, bool) (/bin/clang+0x7abd126)
#12 0x0000558887acab00 llvm::LoopVectorizePass::processLoop(llvm::Loop*) (/bin/clang+0x7acab00)
#13 0x0000558887ad0515 llvm::LoopVectorizePass::runImpl(llvm::Function&) (/bin/clang+0x7ad0515)
#14 0x0000558887ad0a5e llvm::LoopVectorizePass::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/bin/clang+0x7ad0a5e)
#15 0x0000558886dc5f72 llvm::detail::PassModel<llvm::Function, llvm::LoopVectorizePass, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/bin/clang+0x6dc5f72)
#16 0x00005588886d9669 llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/bin/clang+0x86d9669)
#17 0x000055888399a3f2 llvm::detail::PassModel<llvm::Function, llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/bin/clang+0x399a3f2)
#18 0x00005588886dcb10 llvm::ModuleToFunctionPassAdaptor::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/bin/clang+0x86dcb10)
#19 0x00005588839968f2 llvm::detail::PassModel<llvm::Module, llvm::ModuleToFunctionPassAdaptor, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/bin/clang+0x39968f2)
#20 0x00005588886d89e9 llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/bin/clang+0x86d89e9)
#21 0x0000558883993a52 (anonymous namespace)::EmitAssemblyHelper::RunOptimizationPipeline(clang::BackendAction, std::__u::unique_ptr<llvm::raw_pwrite_stream, std::__u::default_delete<llvm::raw_pwrite_stream>>&, std::__u::unique_ptr<llvm::ToolOutputFile, std::__u::default_delete<llvm::ToolOutputFile>>&, clang::BackendConsumer*) (/bin/clang+0x3993a52)
#22 0x000055888398ba80 clang::emitBackendOutput(clang::CompilerInstance&, clang::CodeGenOptions&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::__u::unique_ptr<llvm::raw_pwrite_stream, std::__u::default_delete<llvm::raw_pwrite_stream>>, clang::BackendConsumer*) (/bin/clang+0x398ba80)
#23 0x000055888370719b clang::CodeGenAction::ExecuteAction() (/bin/clang+0x370719b)
#24 0x00005588841b4f41 clang::FrontendAction::Execute() (/bin/clang+0x41b4f41)
#25 0x000055888412b404 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/bin/clang+0x412b404)
#26 0x0000558883701056 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/bin/clang+0x3701056)
#27 0x00005588836f62ce cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/bin/clang+0x36f62ce)
#28 0x00005588836f31c6 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) (/bin/clang+0x36f31c6)
#29 0x00005588842cf51e void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<std::__u::optional<llvm::StringRef>>, std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char>>*, bool*) const::$_0>(long) (/bin/clang+0x42cf51e)
#30 0x000055888885d576 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/bin/clang+0x885d576)
#31 0x00005588842cebfa clang::driver::CC1Command::Execute(llvm::ArrayRef<std::__u::optional<llvm::StringRef>>, std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char>>*, bool*) const (/bin/clang+0x42cebfa)
#32 0x000055888428a066 clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (/bin/clang+0x428a066)
#33 0x000055888428a24f clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::__u::pair<int, clang::driver::Command const*>>&, bool) const (/bin/clang+0x428a24f)
#34 0x00005588842ad530 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::__u::pair<int, clang::driver::Command const*>>&) (/bin/clang+0x42ad530)
#35 0x00005588836f25ac clang_main(int, char**, llvm::ToolContext const&) (/bin/clang+0x36f25ac)
#36 0x00005588836f0c34 main (/bin/clang+0x36f0c34)
#37 0x00007ff9dec2a3d4 __libc_start_main (/lib64/libc.so.6+0x613d4)
#38 0x00005588836f0b6a _start (/bin/clang+0x36f0b6a)
```

Reproducer:  [reduced.ll.zip](https://github.com/user-attachments/files/21166234/reduced.ll.zip)
Repro command:

```
$ clang --target=x86_64-grtev4-linux-gnu \
  -O3 -c -x ir \
  -mllvm -force-vector-interleave=2 \
  -o /dev/null \
  /tmp/reduced.ll
```

Can you please take a look?

david-arm commented:
@bgra8 Sure, I'll have a look!

david-arm commented:
Hi @bgra8, so there is a known issue with forcing vector interleaving with `-mllvm -force-vector-interleave=2`. @fhahn has been working on trying to add support for interleaving with early exit loops. It is currently disabled if you attempt to vectorise normally or using the C level pragma, but you can still force it with `-force-vector-interleave`.

If I remove the `-force-vector-interleave=2` argument it builds fine. I'll create a patch to disable the use of `-force-vector-interleave` with early exit loops for now and we can continue working on it after the release branch point.

bgra8 commented:
Sounds like the two options should both exclude each other?
* disable `early-exit-vectorization` if `force-vector-interleave` is enabled,
* disable `force-vector-interleave` if `early-exit-vectorization` is enabled,
